mm2src/coins/tendermint/tendermint_balance_events.rs

```diff
@@ -0,0 +1,160 @@
+use async_trait::async_trait;
+use common::{executor::{AbortSettings, SpawnAbortable},
+             http_uri_to_ws_address, log};
+use futures_util::{SinkExt, StreamExt};
+use jsonrpc_core::MethodCall;
+use jsonrpc_core::{Id as RpcId, Params as RpcParams, Value as RpcValue, Version as RpcVersion};
+use mm2_core::mm_ctx::MmArc;
+use mm2_event_stream::{behaviour::EventBehaviour, Event, EventStreamConfiguration};
+use mm2_number::BigDecimal;
+use std::collections::{HashMap, HashSet};
+
+use super::TendermintCoin;
+use crate::{tendermint::TendermintCommons, utxo::utxo_common::big_decimal_from_sat_unsigned, MarketCoinOps, MmCoin};
+
+#[async_trait]
+impl EventBehaviour for TendermintCoin {
+    const EVENT_NAME: &'static str = "COIN_BALANCE";
+
+    async fn handle(self, _interval: f64) {
+        fn generate_subscription_query(query_filter: String) -> String {
+            let mut params = serde_json::Map::with_capacity(1);
+            params.insert("query".to_owned(), RpcValue::String(query_filter));
+
+            let q = MethodCall {
+                id: RpcId::Num(0),
+                jsonrpc: Some(RpcVersion::V2),
+                method: "subscribe".to_owned(),
+                params: RpcParams::Map(params),
+            };
+
+            serde_json::to_string(&q).expect("This should never happen")
+        }
+
+        let ctx = MmArc::from_weak(&self.ctx).expect("MM context must have been initialized already.");
```

Comment:
Occasional panic can happen if MM2 is being stopped concurrently with TendermintCoin activation attempt. Also, if we didn't use AbortableSystem, there would be still the dangling MmArc in this endless loop. I recommend calling `from_weak` only when an actual `MmArc` instance is required, and break the loop if `from_weak` fails.

Reply:
> Occasional panic can happen if MM2 is being stopped concurrently with TendermintCoin activation attempt. 

This sounds like a general issue with activations/deactivations. These operation should not be executed in parallel from mm2. Proper locking is needed for handling this sequentially.

> Also, if we didn't use AbortableSystem, there would be still the dangling MmArc in this endless loop. I recommend calling from_weak only when an actual MmArc instance is required, and break the loop if from_weak fails.

We do use abortable system, and the loop is initialized couple lines below of the ctx initialization, so there is no loop unless we upgrade the weak pointer.

Reply:
Just checked the source code, activation and deactivation operations are relying on the `CoinsContext` Mutex.

e.g.,:

`activation requested -> TendermintCoin initialized -> spawning tx/history / balance streaming futures -> locked(no parallel execution allowed here) coins context and added TendermintCoin in there -> success result`

However, spawned futures can end up unexpectedly because there is no guarantee that they will be successfully executed/initialized before we add their coin into `CoinsContext`. This means that a spawned future could encounter errors or panics after we add the coin to the context and return a success result.

I have couple solution ideas on this. Will implement/test them on Monday.